presto-main/src/main/java/io/prestosql/sql/planner/EffectivePredicateExtractor.java

```diff
@@ -272,6 +281,61 @@ public Expression visitJoin(JoinNode node, Void context)
             }
         }
 
+        @Override
+        public Expression visitValues(ValuesNode node, Void context)
+        {
+            if (node.getOutputSymbols().isEmpty()) {
+                return TRUE_LITERAL;
+            }
+
+            ImmutableMap.Builder<Symbol, Domain> domains = ImmutableMap.builder();
+
+            for (int column = 0; column < node.getOutputSymbols().size(); column++) {
+                Symbol symbol = node.getOutputSymbols().get(column);
+                Type type = types.get(symbol);
+
+                ImmutableList.Builder<Object> builder = ImmutableList.builder();
+                boolean hasNull = false;
+                for (int row = 0; row < node.getRows().size(); row++) {
+                    Expression value = node.getRows().get(row).get(column);
+
+                    if (!DeterminismEvaluator.isDeterministic(value)) {
```

Comment:
nit: you could still derive domains for other symbols

Reply:
No, the predicate needs to be an “upper bound”. If we had a single row, we could do that, but otherwise, there’s no way to formulate such a predicate.